File: src/0004_median_of_two_sorted_arrays.py
class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        # assume A=nums1 is shorter than B=nums2
        lenA = len(nums1)
        lenB = len(nums2)
        if lenA > lenB:
            return self.findMedianSortedArrays(nums2, nums1)
        
        lenAB = lenA + lenB
        half = lenAB // 2

        l,r = 0, lenA-1
        while True:
            m = (l+r) // 2
            n = half - m - 2

            Al = nums1[m] if m >= 0 else float("-infinity")
            Ar = nums1[m+1] if m+1 < lenA else float("infinity")
            Bl = nums2[n] if n >= 0 else float("-infinity")
            Br = nums2[n+1] if n+1 < lenB else float("infinity")

            if Al <= Br and Bl <= Ar: # found partition
                if lenAB % 2: 
                    return min(Ar, Br)
                return 0.5*(max(Al, Bl) + min(Ar, Br))

            if Al > Br:
                r = m-1
            else:
                l = m+1
            

File: src/0004_median_of_two_sorted_arrays.cpp
class Solution {
public:
    double findMedianSortedArrays(vector<int>& nums1, vector<int>& nums2) {
        int lenA = nums1.size();
        int lenB = nums2.size();

        if (lenA>lenB)
            return findMedianSortedArrays(nums2, nums1);

        int lenAB = lenA + lenB;
        int half = (lenAB + 1) / 2; // find number of elements in left partition

        int left = 0;
        int right = lenA;
        while (left<=right){
            int Am = (left + right + 1) / 2; 
            int Bm = half - Am;

            int Aleft = (Am>0)?nums1[Am-1]:INT_MIN;
            int Aright = (Am<lenA)?nums1[Am]:INT_MAX;
            int Bleft = (Bm>0)?nums2[Bm-1]:INT_MIN;
            int Bright = (Bm<lenB)?nums2[Bm]:INT_MAX;
            if ((Aleft <= Bright) && (Bleft <= Aright)) {
                if (lenAB % 2 == 1)
                    return ((double) max(Aleft, Bleft));
                return ((double) 0.5*(max(Aleft, Bleft) + min(Aright, Bright)));
            }
            if (Aleft > Bright)
                right = Am - 1;
            else
                left = Am + 1;
        }
        return -1;
    }
};